hts.c

```diff
@@ -438,6 +444,13 @@ htsFile *hts_open_format(const char *fn, const char *mode, const htsFormat *fmt)
     fp = hts_hopen(hfile, fn, smode);
     if (fp == NULL) goto error;
 
+    // htsExactFormat goes from a nice exact input format (bam, sam)
+    // to a woefully inexact one (binary, text).  Honour the original one.
+    if (fp->is_write && fmt &&
+        (fmt->format == bam || fmt->format == sam ||
+         fmt->format == vcf || fmt->format == bcf))
+        fp->format.format = fmt->format;
+
```

Comment:
>  // htsExactFormat goes from a nice exact input format (bam, sam)
>  // to a woefully inexact one (binary, text).  Honour the original one.

It'd be great to do this without the editorialising. The reason for `binary_format`/`text_format` has been explained before, e.g., in https://github.com/samtools/htslib/pull/721#issuecomment-400666150.

What's really going on here is the exactness is lost because `smode` (given to `hts_hopen()`) is an approximation of the format. Ideally this would be solved with some refactoring so no approximating was necessary; but more simple would be to track the real exactFormat on line 424 above (alongside `*mode_c = format_to_mode[fmt->format]`) and set `fp->format.format` accordingly here `if (fp->is_write && that_code_path_was_taken)`.

That would avoid this test for `bam`/`sam`/`vcf`/`bcf` and so make this work for e.g. `bed` as well.

Reply:
I'll rewrite the comment and add bed, although I don't think it's editorial - it's 100% factual.  We can't really get rid of the rather fragile set of formats until we fix other things, so this is a pragmatic workaround for now.  I deliberately avoided copying over format for everything beyond `unknown_format` because I confess I didn't really understand the implications of types on things like json or the index formats.  It's rather twisty logic in places, so played it safe with the bits I can actually test don't cause breakage.

If you think it's safe we could get rid of the whole format checking bit and just use `&& fmt->format > unknown_format`, but I don't think our test harness would cover a lot of it as it doesn't have web stuff in there.

Regarding PR #721, the last we heard was you were going to rework it.  Did that get shelved?  It'd definitely be great to have this tidied up.

I also see we stated "the release candidate is meant to be today" in June 2018 and the new stuff would arrive asap afterwards. Um... no comment!

Reply:
“nice” and “woefully” are hardly 100% factual.

Yes, PR #721 is back on my to-do list to bring up to date during the current apparent HTSlib frenzy :smile:.

IIRC the web stuff only occurs during reading, so fixups that occur for is_write are fine for web stuff. Point is: for writing, `fp->format.format` should end up being either exactly the same as or compatible with the `fmt->format` argument, otherwise there is a bug in the `smode` shoehorning or in `hts_hopen()`. So if line 424 `*mode_c = …` was triggered and we're writing, then it is safe to set `fp->format.format` to the same thing that `*mode_c` was based on, i.e., the input `fmt->format` argument.

I think a big part of the problem here is that it's unclear how `hts_open_format`'s `mode` and `fmt` arguments are supposed to overrule each other. (The documentation just says TODO…)

Most sensible might be to have `hts_open_format(fn, mode, NULL)` call `hts_open(fn, mode)` effectively; otherwise when `fmt` is non-NULL `mode` is just examined for `r` or `w` and otherwise ignored. But I haven't worked that through…